### PR TITLE
docs: Add tutorial for configuring Dovecot FTS with Solr

### PR DESCRIPTION
**Pull request by beertje44**
# Description

Tutorial on using standard apache solr container with dovecot provided solr schema as dovecot fts backend in dms.

ref: https://github.com/docker-mailserver/docker-mailserver/pull/4025

## Type of change

- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary, I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] **I have added information about changes made in this PR to `CHANGELOG.md`**

### CODE REVIEW

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+*Dovecot full text search using solr backend*
+
+Dovecot can use several fts backends to support efficient and fast full text searching of e-mails directly from the imap server. Especially if you have one or more large mail folders this can make a huge difference, since the alternative is dovecot searching through each and every email all by itself on the spot, agian and again. The latter most times means one cannot search through a large pile off emails with clients that don't store all imap mail locally, for example mobile clients like Gmail or webmail clients.
+One of these is Apache SOLR, an fast and efficient multi-purpose search indexer.
+
+Firstly you need a working solr container, for this the official docker container will do:
+
+```
+---
+version: "3.3"
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:10]: `version` is deprecated.
```suggestion

```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+*Dovecot full text search using solr backend*
+
+Dovecot can use several fts backends to support efficient and fast full text searching of e-mails directly from the imap server. Especially if you have one or more large mail folders this can make a huge difference, since the alternative is dovecot searching through each and every email all by itself on the spot, agian and again. The latter most times means one cannot search through a large pile off emails with clients that don't store all imap mail locally, for example mobile clients like Gmail or webmail clients.
+One of these is Apache SOLR, an fast and efficient multi-purpose search indexer.
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:4]: Not 100% sure about this one:

```suggestion
One of these is Apache SOLR, a fast and efficient multi-purpose search indexer.
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+Firstly you need a working solr container, for this the official docker container will do:
+
+```
+---
+
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:10]:
```suggestion
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  solr:
+    image: solr:latest
+    environment:
+      SOLR_JAVA_MEM: "-Xms2g -Xmx2g"
+    volumes:
+      - <local folder>:/var/solr
+    restart: always
+```
+
+We'll assume dms will connect internally to solr, so either append the above docker composer snippet to your dms compose.yml or make sure both containers use the same docker network.
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:21]:
```suggestion
We'll assume dms will connect internally to solr, so either append the above docker composer snippet to your dms `compose.yml` or make sure both containers use the same docker network.
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+- solr-config-9.xml
+- solr-schema-9.xml
+
+Copy solr-config-9.xml to the data/dovecot folder and name it: solrconfig.xml
+Copy solr-schema-9.xml to the data/dovecot folder and name it: schema.xml, remove managed-schema.xml
+Both files should be owned by uid and gid 8983.
+
+Start the solr container once again, you should now have a working dovecot fts specific solr core. All that is left is to connect dms dovecot to this solr core:
+
+Create a 10-plugin.conf file in your config/dovecot folder and link it in your compose.yml like so:
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:44]:
```suggestion
Create a `10-plugin.conf` file in your `config/dovecot` folder and link it in your `compose.yml` like so:
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+solr create -c dovecot
+cp -R /opt/solr/contrib/analysis-extras/lib /var/solr/data/dovecot
+```
+
+Stop the container, you should now have a data/dovecot folder. All that is needed on the solr part is a schema that is tailored specifically for dovecot fts. Luckally dovecot provides these:
+https://github.com/dovecot/core/tree/main/doc
+
+As of writing of this guide solr 9 is current, so you need the 2 solr 9 config files:
+- solr-config-9.xml
+- solr-schema-9.xml
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:36]:
```suggestion
- `solr-config-9.xml`
- `solr-schema-9.xml`
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+Copy solr-config-9.xml to the data/dovecot folder and name it: solrconfig.xml
+Copy solr-schema-9.xml to the data/dovecot folder and name it: schema.xml, remove managed-schema.xml
```
Comment (casperklein) [inline docs/content/examples/tutorials/dovecot-solr.md:39]:
```suggestion
Copy `solr-config-9.xml` to the `data/dovecot` folder and name it: `solrconfig.xml`
Copy `solr-schema-9.xml` to the `data/dovecot` folder and name it: `schema.xml`, remove `managed-schema.xml`
```

**Comment by casperklein** [comment]
```
fts      --> FTS
imap     --> IMAP
dms      --> DMS
dovecot  --> Dovecot
solr     --> Solr 
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
+*Dovecot full text search using solr backend*
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:1]:
```suggestion
# Dovecot Full Text Search (FTS) using the Solr Backend
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+Dovecot can use several fts backends to support efficient and fast full text searching of e-mails directly from the imap server. Especially if you have one or more large mail folders this can make a huge difference, since the alternative is dovecot searching through each and every email all by itself on the spot, agian and again. The latter most times means one cannot search through a large pile off emails with clients that don't store all imap mail locally, for example mobile clients like Gmail or webmail clients.
+One of these is Apache SOLR, an fast and efficient multi-purpose search indexer.
+
+Firstly you need a working solr container, for this the official docker container will do:
+
+```
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:8]: Please add `yaml` after the 3 backticks so we get syntax highlighting.

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    volumes:
+      - <local folder>:/var/solr
+    restart: always
+```
+
+We'll assume dms will connect internally to solr, so either append the above docker composer snippet to your dms compose.yml or make sure both containers use the same docker network.
+The enviroment setting SOLR_JAVA_MEM is optional, but solr can be quite resource hungry so the default of 512MB can be exhausted rather quickly.
+
+Once started you need to configure a solr core for dovecot:
+```
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:25]: Add a newline between these two and use the `bach` syntax highlighter (write `bash` after the three backticks).

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+docker exec -it solr_solr_1 /bin/sh
+solr create -c dovecot
+cp -R /opt/solr/contrib/analysis-extras/lib /var/solr/data/dovecot
+```
+
+Stop the container, you should now have a data/dovecot folder. All that is needed on the solr part is a schema that is tailored specifically for dovecot fts. Luckally dovecot provides these:
+https://github.com/dovecot/core/tree/main/doc
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:32]:
```suggestion
Stop the container, you should now have a data/dovecot folder. All that is needed on the solr part is a schema that is tailored specifically for dovecot fts. [Luckilly, Dovecot provides these](https://github.com/dovecot/core/tree/main/doc).
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+As of writing of this guide solr 9 is current, so you need the 2 solr 9 config files:
+- solr-config-9.xml
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:35]:
```suggestion
As of writing of this guide solr 9 is current, so you need the 2 solr 9 config files:

- `solr-config-9.xml`
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+- solr-schema-9.xml
+
+Copy solr-config-9.xml to the data/dovecot folder and name it: solrconfig.xml
+Copy solr-schema-9.xml to the data/dovecot folder and name it: schema.xml, remove managed-schema.xml
+Both files should be owned by uid and gid 8983.
+
+Start the solr container once again, you should now have a working dovecot fts specific solr core. All that is left is to connect dms dovecot to this solr core:
+
+Create a 10-plugin.conf file in your config/dovecot folder and link it in your compose.yml like so:
+```
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:45]: Add `yaml` after the backticks.

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+volumes:
+  ...
+  <local dms folder>/config/dovecot/10-plugin.conf:/etc/dovecot/conf.d/10-plugin.conf:ro
+  ...
+```
+
+It's content should be:
+```
```
Comment (georglauterbach) [inline docs/content/examples/tutorials/dovecot-solr.md:53]: Newline + `yaml` after backticks.

**Review by georglauterbach — CHANGES_REQUESTED** [review]
After @casperklein's and my feedback has been addressed, you will still need to adjust the `mkdocs.yml` to integrate your new page into our documentation.

The PR itself is still in a rough shape; I will also request @polarathene for a final review, as he has the most experience with our docs. I don't want him to have a third look currently, though; having the last word and final review is appropriate IMO.

**Comment by georglauterbach** [comment]
@polarathene go ahead and have a look :) I suggest you immediately apply feedback that you have 🚀 

---

@beertje44 you can now already take a look at the preview we generate. Look out for comments from the github-actions bot, [like this one](https://github.com/docker-mailserver/docker-mailserver/pull/4070#issuecomment-2171796039).

**Comment by polarathene** [comment]
Current preview of new docs page: https://pullrequest-4070--dms-doc-previews.netlify.app/examples/tutorials/dovecot-solr/

I have revised the PR a bit:

- There was some parts from earlier review that had not been addressed (_notably naming convention from_ @casperklein and new line requirements from @georglauterbach ) , additionally some typos to correct.
- Other portions have been revised to be worded better, or restructured for better formatting for the reader to digest.
- Relevant links added for context. Especially the UID/GID reference.

---

There is no mention for the `dovecot-solr` package requirement. My understanding is that we are not bringing that back as it's missing support for ARM64 releases?

I'll add a new commit to advise on including the package and ARM64 caveat, then this should be good for merging.

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+```bash
+docker compose exec mailserver doveadm fts rescan -A
+```
+
+!!! info "Indexing will take a while depending on how large your mail folders"
+
+    Usually within 15 minutes or so, you should be able to search your mail using the Dovecot FTS feature! :tada:
+
+[docs::user-patches]: ../config/advanced/override-defaults/user-patches.md
```
Comment (polarathene) [inline docs/content/examples/tutorials/dovecot-solr.md:139]:
```suggestion
[docs::user-patches]: ../../config/advanced/override-defaults/user-patches.md
[docs::dovecot::full-text-search]: ../../config/advanced/full-text-search.md
```

File: docs/mkdocs.yml [hunk 16]
```diff
@@ -174,6 +174,7 @@ nav:
       - 'Crowdsec': examples/tutorials/crowdsec.md
       - 'Building your own Docker image': examples/tutorials/docker-build.md
       - 'Blog Posts': examples/tutorials/blog-posts.md
+      - 'Full text search using Apache SOLR': examples/tutorials/dovecot-solr.md
```
Comment (polarathene) [inline docs/mkdocs.yml:177]:
```suggestion
      - 'Dovecot FTS with Apache Solr': examples/tutorials/dovecot-solr.md
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 17]
```diff
@@ -0,0 +1,145 @@
+# Dovecot Full Text Search (FTS) using the Solr Backend
+
+Dovecot supports several FTS backends for providing fast and efficient full text searching of e-mails directly from the IMAP server.
+
+As the size of your mail storage grows, the benefits of FTS is especially notable:
```
Comment (polarathene) [inline docs/content/examples/tutorials/dovecot-solr.md:5]:
```suggestion
Dovecot supports several FTS backends for providing fast and efficient full text searching of e-mails directly from the IMAP server.

As the size of your mail storage grows, the benefits of FTS are especially notable:
```

File: docs/content/examples/tutorials/dovecot-solr.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+- Without FTS, Dovecot would perform a search query by checking each individual email stored for a match, and then repeat this process again from scratch for the exact same query in future.
+- Some mail clients (_like Thunderbird_) may provide their own indexing and search features when all mail to search is stored locally, otherwise Dovecot needs to handle the search query (_for example webmail and mobile clients, like Gmail_).
+- FTS indexes each mail into a database for querying instead, where it can skip the cost of inspecting irrelevant emails for a query.
+
```
Comment (polarathene) [inline docs/content/examples/tutorials/dovecot-solr.md:10]:
```suggestion

!!! warning "This is a community contributed guide"

    It extends [our official docs for Dovecot FTS][docs::dovecot::full-text-search] with a focus on Apache Solr. DMS does not officially support this integration.

```

**Review by polarathene — CHANGES_REQUESTED** [review]
I'll apply these.

**Review by polarathene — APPROVED** [review]

**Comment by polarathene** [comment]
Thanks everyone!

This guide seems to be missing some advice that was detailed by @erm967 in https://github.com/docker-mailserver/docker-mailserver/issues/4052#issue-2338715252 , is that not required for some reason?

I suspect that is one of the maintenance caveats to keep in mind as the external resource won't be tied to a specific DMS release (_at least if this new doc page is not actively maintained across releases_).

---

I've left the referenced feature request open, but I've also included a link on the new docs page for any interested users to express their interest in the package. I think we've covered why we no longer include the package and the workaround for Solr users is quite simple as covered in the docs.

Closing that feature request with this PR is acceptable. It can always be opened again in future if relevant.

**Review by georglauterbach — APPROVED** [review]

**Pull request merged**